### PR TITLE
Run Control Plane components as `nonroot` user `65532`

### PR DESCRIPTION
**Pull request by AleksandarSavchev**
**How to categorize this PR?**

/area security quality
/kind enhancement

**What this PR does / why we need it**:
This PR modifies Control Plane components `kube-apiserver`, `kube-controller-manager` and `kube-scheduler` to run as `nonroot` user `65532` (id of `nonroot` user in `distroless` images [ref](https://github.com/GoogleContainerTools/distroless/blob/main/base/base.bzl#L8)). The aim of this PR is to adhere to the principles of least privilege and reduce potential attack surface.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```other operator
Control plane components `kube-apiserver`, `kube-controller-manager` and `kube-scheduler` now run as `nonroot` user and group `65532`.
```

### CODE REVIEW

**Review by dimityrmirchev — APPROVED** [review]
/lgtm

File: pkg/component/kubeapiserver/deployment.go [hunk 1]
```diff
@@ -233,6 +233,10 @@ func (k *kubeAPIServer) reconcileDeployment(
 						Args:                     k.computeKubeAPIServerArgs(),
 						TerminationMessagePath:   corev1.TerminationMessagePathDefault,
 						TerminationMessagePolicy: corev1.TerminationMessageReadFile,
+						SecurityContext: &corev1.SecurityContext{
+							RunAsNonRoot: pointer.Bool(true),
+							RunAsUser:    pointer.Int64(65532),
```
Comment (ialidzhikov) [inline pkg/component/kubeapiserver/deployment.go]: If the upstream images can run with this `65532` why they run as root? I think we should open upstream issue to ask also the K8s folks to change the default user for the container image to `65532`.

**Comment by AleksandarSavchev** [comment]
Converted to draft in order to research setting `runAsGroup: 65532` as well.

File: pkg/component/kubeapiserver/deployment.go (hunk 1 above)
Comment (AleksandarSavchev) [reply pkg/component/kubeapiserver/deployment.go]: They can run when set to run as `65532`, however there can be a problem with the volumes and which files are configured to be used from them.

Currently all our secret and configMap volumes have set `defaultMode: 420` which gives `644` permission to these files. Since these files are owned by `root` setting `640` would remove any permissions to the `65532` user.

I have now moved the settings to the podSecurityContext, where I can set `fsGroup` to `65532` which would change the group owner of these files to `65532` and we can lower the permissions of mounted files to `640`. There are `*.key` files which we would ideally want to have `600` permission, but currently it is not possible to change the user owner of files [ref](https://github.com/kubernetes/kubernetes/issues/81089)

In conclusion, I do not think it is needed to create such issue, since the change would be breaking and may users may have opted to use `defaultMode: 384` which would give `600` permissions.

**Review by ScheererJ — COMMENTED** [review]
Did you check with HA vpn? In the HA-mode of vpn, there are vpn side-car containers added to `kube-apiserver`. Do they also work with this user change?

**Comment by AleksandarSavchev** [comment]
> Did you check with HA vpn? In the HA-mode of vpn, there are vpn side-car containers added to `kube-apiserver`. Do they also work with this user change?

I have not checked the HA scenarios. I will look into them, if the vpn side-car containers are problematic the easiest fix would be to set their container to run as `root`.

**Comment by AleksandarSavchev** [comment]
After validating locally the `vpn` containers cannot run as nonroot and would error with the following: `panic: open /tmp/acquired-ip: permission denied`.
I tested adding the `nonroot` user to the image used in [this Dockerfile ](https://github.com/gardener/vpn2/blob/9c368036ad3e102f28a62d49bfc4a688744972af/shoot-client/Dockerfile) and got the following error: `/network-connection.sh: line 47: /proc/sys/net/ipv4/tcp_keepalive_time: Permission denied` [ref](https://github.com/gardener/vpn2/blob/9c368036ad3e102f28a62d49bfc4a688744972af/shoot-client/network-connection.sh#L51-L57). The `/proc/sys/net/ipv4` files have `0644` permissions and their owners cannot be changed from `root`.
I have decided to leave the `vpn` containers running as `root`.

File: pkg/component/kubeapiserver/deployment.go [hunk 2]
```diff
@@ -225,6 +225,12 @@ func (k *kubeAPIServer) reconcileDeployment(
 					RestartPolicy:                 corev1.RestartPolicyAlways,
 					SchedulerName:                 corev1.DefaultSchedulerName,
 					TerminationGracePeriodSeconds: pointer.Int64(30),
+					SecurityContext: &corev1.PodSecurityContext{
+						RunAsNonRoot: pointer.Bool(true),
+						RunAsUser:    pointer.Int64(65532),
+						RunAsGroup:   pointer.Int64(65532),
+						FSGroup:      pointer.Int64(65532),
```
Comment (dimityrmirchev) [inline pkg/component/kubeapiserver/deployment.go:234]: I would suggest that we add a comment so no one wonders where this user comes from. The same can be done for kube-scheduler and KCM.
```suggestion
                                                  // use the nonroot user from a distroless container
                                                  // https://github.com/GoogleContainerTools/distroless/blob/1a8918fcaa7313fd02ae08089a57a701faea999c/base/base.bzl#L8
						RunAsNonRoot: pointer.Bool(true),
						RunAsUser:    pointer.Int64(65532),
						RunAsGroup:   pointer.Int64(65532),
						FSGroup:      pointer.Int64(65532),
```

**Review by dimityrmirchev — APPROVED** [review]
/lgtm

File: pkg/component/kubeapiserver/deployment.go [hunk 3]
```diff
@@ -828,6 +836,8 @@ func (k *kubeAPIServer) vpnSeedClientContainer(index int) *corev1.Container {
 			},
 		},
 		SecurityContext: &corev1.SecurityContext{
+			RunAsNonRoot: pointer.Bool(false),
+			RunAsUser:    pointer.Int64(0),
```
Comment (ialidzhikov) [inline pkg/component/kubeapiserver/deployment.go:840]: What happens when a Pod has a securityContext set at Pod level and then some container overwrite it? Does the container securityContext is considered only or does kubernetes internally merges the container securityContext with the pod securityContext (where container securityContext values take precedence)? 
I am asking because if a merge happens under the hood, then the merged securityContext would be
```yaml
securityContext:
  runAsNonRoot: false
  runAsUser: 0
  runAsGroup: 65532
  fsGroup: 65532
```

which I think is not what we want.
The docs write 
>  Security settings that you specify for a Container apply only to the individual Container, and they override settings made at the Pod level when there is overlap.

but I am not 100% sure.

Comment (ialidzhikov) [reply pkg/component/kubeapiserver/deployment.go:840]: It seems that a merge happens under the hood indeed. 

A small isolated setup to prove the merge:
1. Create the following yaml:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: alpine
  namespace: default
spec:
  securityContext:
    runAsNonRoot: true
    runAsUser: 65532
    runAsGroup: 65532
    fsGroup: 65532
  containers:
  - image: alpine:3.13.2
    command:
      - /bin/sh
      - "-c"
      - "sleep 60m"
    imagePullPolicy: IfNotPresent
    name: alpine
    securityContext:
      runAsNonRoot: false
      runAsUser: 0
  restartPolicy: Always
```

2. Verify the merge

```
% k exec -ti alpine -- sh
/ # id
uid=0(root) gid=65532 groups=1(bin),2(daemon),3(sys),4(adm),6(disk),10(wheel),11(floppy),20(dialout),26(tape),27(video),65532
```

Again, maybe this is not what we want? Maybe it is better to move the Pod level securityContext to the kube-apiserver container?

Comment (AleksandarSavchev) [reply pkg/component/kubeapiserver/deployment.go:840]: Yes, the Pod's `securityContext` setting will be used by every container and only the specified settings will be overridden by the container's `securityContext`. It is important to note that `fsGroup` can be set only at the Pod level settings so the only question is if we should explicitly set `runAsGroup` at the Container level.

I do not think that it is required since we already set `root` as the user which grants full permissions on all files in the container.

**Review by ialidzhikov — APPROVED** [review]
